ETE Tests of TimeLock Migration [TimeLock ETE Tests, Part III]

jeremyk-91 opened the pull request:
**Goals (and why)**:

* Add end-to-end tests for automated migrations of the AtlasDB TimeLock Server. This allows us to more easily check that automated migrations behave as we expect them to. There are already tests for major pieces of the puzzle (TimeLockMigrator, CassandraTimestampInvalidatorStore), but we don't have a full end-to-end workflow.

**Implementation Description (bullets)**:

* `TimeLockMigrationEteTest` executes the following workflow:
  1. sets up three containers: a Cassandra, a TimeLock, and an AtlasDbEteServer (an Atlas client) configured with embedded services
  2. executes a write and a read on the Atlas client (`"contains one todo pre-migration"`)
  3. gets a timestamp `TS` from the Atlas client
  4. reconfigures the Atlas client to use TimeLock and bounces it
  11. gets a timestamp directly from TimeLock `TS'`, verifies `TS' > TS` (`"timestamp was migrated to TimeLock"`)
  5. executes a read on the Atlas client (`"can still read after migration to TimeLock"`)
  6. executes a write and a read on the Atlas client (`"can add a new todo using TimeLock"`)
  7. attempts to get a fresh timestamp from the Atlas client, expecting failure (`"no longer exposes an embedded timestamp service"`)
  8. downgrades the Atlas client to embedded services, without doing any kind of reverse migration, and bounces it
  9. checks that the Atlas client can't create a transaction manager by parsing its logs
  10. attempts to read/write todos from the Atlas client, expecting failure (`"cannot read/write using embedded service after migration to TimeLock"`)

* `DockerClientOrchestrationRule` is the underlying glue to make things work, setting up/tearing down the various containers as well as executing code, pulling down logs, etc.

**Concerns (what feedback would you like?)**:

* `DockerClientOrchestrationRule` was pretty painful to write (basically, you cannot just use a `RuleChain` to setup the DCR and DPR, since they can only be created once the TemporaryFolder's location is known - and that only happens once its `before()` method executes). I hope it's not too unreadable.
* There is some duplication with `EteSetup` which I couldn't use directly because I needed better control of the orchestration.

**Where should we start reviewing?**:

* `TimeLockMigrationEteTest` is probably the best start point.

**Priority (whenever / two weeks / yesterday)**:

* Next week please. Not massively urgent.

jeremyk-91 commented:
There's a small amount of polish I want to add to this.

harrybiddle commented on atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/DockerClientOrchestrationRule.java:
```diff
@@ -0,0 +1,153 @@
+                .build();
+    }
+
+    private String dockerExecOnClient(String... arguments) {
```
Could this be shared with the e.g. nodetool repair and other cases where we use the ete container?

jeremyk-91 replied:
I'm not sure we actually run anything directly on the boxes (other than `ete-cli`) in the rest of the ETE tests. We could possibly unify those two (a static taking `DockerComposeRule`, a container and string arguments), though I'm not sure I want to touch the rest of the ETE tests in this PR.

harrybiddle commented on atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/DockerClientOrchestrationRule.java:
```diff
@@ -0,0 +1,153 @@
+    }
+
+    public String getClientLogs() {
+        return dockerExecOnClient("cat", "var/log/atlasdb-ete-startup.log");
```
if the docker command was ``init.sh && tail -f var/log/atlasdb-ete-startup.log``, this should be just grabbed by docker-compose-rule. Maybe this doesn't work in this case since the ETE container is so general, just a thought...

jeremyk-91 replied:
Hmm, I seemed to have some difficulty with this because DockerComposeRule's `writeLogs` blocks until the container is done. I could inject an InterruptedException, perhaps, but I think `cat` will be less hacky here...

jeremyk-91 commented:
Thanks @harrybiddle! I'll get to the comments, but some preliminary thoughts overall:

• I think we should still keep the `DropwizardEteTest`, because it tests that Console and the CLIs (via the Dropwizard bundle) work properly with TimeLock. This is used e.g. by internal backup/restore product. 
• But `DropwizardEteTest` is now probably a poor choice of name, given that I've extracted the `exposesTimestampService()` test out to its own class. Maybe something like `CliEteTest` might work better as a name?
• The failure in 2237 is caused by `cannot verify github-cloud.s3.amazonaws.com's certificate` in docker-compose, it seems, so I wouldn't be concerned about it. Thanks for running the tests!

jeremyk-91 commented on atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/TimeLockMigrationEteTest.java:
```diff
@@ -0,0 +1,183 @@
+// ETE tests where the general idea is "set up all the containers, and fire".
+public class TimeLockMigrationEteTest {
+    // Docker Engine daemon only has limited access to the filesystem, if the user is using Docker-Machine
+    // Thus root the temporary folder as a subdirectory of the user's home directory
```
Ugh, not sure why I wrote that that way. I simply meant to place / store it there...

jeremyk-91 replied:
Changed to "// Thus ensure the temporary folder is a subdirectory of the user's home directory"

jeremyk-91 commented on atlasdb-ete-tests/src/test/java/com/palantir/atlasdb/ete/TimeLockMigrationEteTest.java:
```diff
@@ -0,0 +1,183 @@
+        waitUntil(serversAreReady());
+    }
+
+    private void downgradeAtlasClientFromTimelock() {
```
Agreed, added "withoutMigration" at the end.

jeremyk-91 commented:
In addition to the comments:
• renamed `DropwizardEteTest` to `CommandLineEteTest`
• changed the docker compose file to run `tail -f var/log/atlasdb-ete-startup.log` rather than `sleep 99999999` at the end (giving us better logs in the artifacts as well).

harrybiddle commented:
Looks good! Happy to merge this.
